packages/core/src/shared/telemetry/util.ts

```diff
@@ -275,3 +277,54 @@ export function getOperatingSystem(): 'MAC' | 'WINDOWS' | 'LINUX' {
         return 'LINUX'
     }
 }
+
+/**
+ * Decorator that simply wraps the method with a non-emitting telemetry `run()`, automatically
+ * `record()`ing the provided function id for later use by {@link TelemetryTracer.getFunctionStack()}
+ *
+ * This saves us from needing to wrap the entire function:
+ *
+ * **Before:**
+ * ```
+ * class A {
+ *     myMethod() {
+ *         telemetry.function_call.run(() => {
+ *                 ...
+ *             },
+ *             { emit: false, functionId: { name: 'myMethod', class: 'A' } }
+ *         )
+ *     }
+ * }
+ * ```
+ *
+ * **After:**
+ * ```
+ * class A {
+ *     @withTelemetryContext({ name: 'myMethod', class: 'A' })
+ *     myMethod() {
+ *         ...
+ *     }
+ * }
+ * ```
+ */
+export function withTelemetryContext(functionId: FunctionEntry) {
```

Comment:
I had it in an early comment but adding emit as an optional arg here might save you from having to indent a few more methods

Reply:
Not completely sure what you mean by this. Did you have an example

Reply:
I think it was only relevant to https://github.com/aws/aws-toolkit-vscode/pull/5465#discussion_r1723535776 if you wanted to change emit to true/false through the decorator. Looking at it again I don't think its needed though